### PR TITLE
WIP: Introduce payload primitives

### PR DESCRIPTION
**Pull request by ghost**
WIP
Ref https://github.com/paradigmxyz/reth/issues/8330

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -34,6 +34,7 @@ members = [
     "crates/payload/builder/",
     "crates/payload/ethereum/",
     "crates/payload/validator/",
+    "crates/payload-primitives/",
```
Comment (mattsse) [inline Cargo.toml:37]: this shouled be moved into the payload folder

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, like the direction

**Comment by 0xqd** [comment]
@mattsse I'm looking into moving this PayloadAttributes trait from engine-primitives to payload-primitives but it has 1 function which belongs to engine which is ensure_well_form_attributes()

```rust
/// The execution payload attribute type the CL node emits via the engine API.
/// This trait should be implemented by types that could be used to spawn a payload job.
///
/// This type is emitted as part of the forkchoiceUpdated call
pub trait PayloadAttributes:
    serde::de::DeserializeOwned + serde::Serialize + std::fmt::Debug + Clone + Send + Sync + 'static
{
    /// Returns the timestamp to be used in the payload job.
    fn timestamp(&self) -> u64;

    /// Returns the withdrawals for the given payload attributes.
    fn withdrawals(&self) -> Option<&Vec<Withdrawal>>;

    /// Return the parent beacon block root for the payload attributes.
    fn parent_beacon_block_root(&self) -> Option<B256>;

    /// Ensures that the payload attributes are valid for the given [ChainSpec] and
    /// [EngineApiMessageVersion].
    fn ensure_well_formed_attributes(
        &self,
        chain_spec: &ChainSpec,
        version: EngineApiMessageVersion,
    ) -> Result<(), EngineObjectValidationError>;
}
```

Should I create another trait called ValidatePayloadAttributes in engine, and implement it. And remove the function in PayloadAttribute ?

File: Cargo.toml [hunk 2]
```diff
@@ -47,6 +47,7 @@ members = [
     "crates/payload/builder/",
     "crates/payload/ethereum/",
     "crates/payload/validator/",
+    "crates/payload/primitives/",
```
Comment (mattsse) [inline Cargo.toml:50]: swap position with validator

File: Cargo.toml [hunk 3]
```diff
@@ -265,6 +266,7 @@ reth-optimism-consensus = { path = "crates/optimism/consensus" }
 reth-optimism-payload-builder = { path = "crates/optimism/payload" }
 reth-payload-builder = { path = "crates/payload/builder" }
 reth-payload-validator = { path = "crates/payload/validator" }
+reth-payload-primitives = { path = "crates/payload/primitives" }
```
Comment (mattsse) [inline Cargo.toml:269]: same

File: bin/reth/Cargo.toml [hunk 4]
```diff
@@ -42,6 +42,7 @@ reth-tasks.workspace = true
 reth-ethereum-payload-builder.workspace = true
 reth-payload-builder.workspace = true
 reth-payload-validator.workspace = true
+reth-payload-primitives.workspace = true
```
Comment (mattsse) [inline bin/reth/Cargo.toml:45]: same

File: crates/payload/primitives/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+homepage.workspace = true
+repository.workspace = true
+
+[lints]
+workspace = true
+
+[dependencies]
+# reth
+reth-provider.workspace = true
+reth-interfaces.workspace = true
```
Comment (mattsse) [inline crates/payload/primitives/Cargo.toml:16]: can we do without these here?

File: crates/payload/primitives/src/payload_or_attribute.rs
Comment (mattsse) [inline crates/payload/primitives/src/payload_or_attribute.rs:1]: I think this does not belong in here because this an engine thing

**Review by mattsse — CHANGES_REQUESTED** [review]
quite happy with the structure, some nits

> Should I create another trait called ValidatePayloadAttributes in engine

yes, we want exactly this!
this way payload(attributes) is entirely decoupled from the engine

**Comment by mattsse** [comment]
sorry about the conflicts -.-

checking in the am

ty 🫡

**Comment by 0xqd** [comment]
> sorry about the conflicts -.-
> 
> checking in the am
> 
> ty 🫡

no worry, I need to double check on the EngineTypes as well, since we support V4.

**Review by mattsse — CHANGES_REQUESTED** [review]
hmm, unclear where the conflicts are coming from tbh

can't really review this, but the payload changes are looking good

File: crates/payload/basic/src/lib.rs [hunk 6]
```diff
@@ -11,10 +11,11 @@
 use crate::metrics::PayloadBuilderMetrics;
 use futures_core::ready;
 use futures_util::FutureExt;
-use reth_engine_primitives::{BuiltPayload, PayloadBuilderAttributes};
-use reth_payload_builder::{
-    database::CachedReads, error::PayloadBuilderError, KeepPayloadJobAlive, PayloadId, PayloadJob,
-    PayloadJobGenerator,
+
```
Comment (mattsse) [inline crates/payload/basic/src/lib.rs:14]:
```suggestion
```

File: examples/custom-payload-builder/src/generator.rs [hunk 7]
```diff
@@ -5,8 +5,9 @@ use reth::{
     transaction_pool::TransactionPool,
 };
 use reth_basic_payload_builder::{BasicPayloadJobGeneratorConfig, PayloadBuilder, PayloadConfig};
-use reth_node_api::PayloadBuilderAttributes;
-use reth_payload_builder::{error::PayloadBuilderError, PayloadJobGenerator};
+use reth_payload_primitives::{
+    error::PayloadBuilderError, PayloadBuilderAttributes, PayloadJobGenerator,
+};
```
Comment (mattsse) [inline examples/custom-payload-builder/src/generator.rs:10]: all of them should be re-exported form reth-basic-payload-builder, so we don't need the additional dep

File: crates/engine-primitives/src/lib.rs [hunk 8]
```diff
@@ -64,312 +58,312 @@ pub trait EngineTypes:
...
-    #[test]
-    fn version_ord() {
-        assert!(EngineApiMessageVersion::V4 > EngineApiMessageVersion::V3);
-    }
-}
+// /// Validates the timestamp depending on the version called:
+// ///
+// /// * If V2, this ensures that the payload timestamp is pre-Cancun.
+// /// * If V3, this ensures that the payload timestamp is within the Cancun timestamp.
+// /// * If V4, this ensures that the payload timestamp is within the Prague timestamp.
```
Comment (mattsse) [inline crates/engine-primitives/src/lib.rs:65]: can be removed?

**Review by mattsse — CHANGES_REQUESTED** [review]
we still need to get rid of this:

https://github.com/paradigmxyz/reth/blob/main/crates/payload/builder/src/service.rs#L35-L35

but we can do this separately as this pr is already very invasive.

File: crates/engine-primitives/src/lib.rs (hunk 8 above)
Comment (0xqd) [reply crates/engine-primitives/src/lib.rs:65]: yes, I'm working on another branch. based on current changes, will force push to this one or create another PR. Thanks for your feedback.

**Comment by 0xqd** [comment]
I will close this PR, and open new one to avoid solving conflicts. All changes are tracked.

**Pull request closed without merging**